src/Storages/Rocksdb/StorageEmbeddedRocksdb.cpp

```diff
@@ -0,0 +1,318 @@
+{
+    const auto & select = query->as<ASTSelectQuery &>();
+    if (!select.where())
+    {
+        return std::make_pair(FieldVector{}, true);
+    }
+    FieldVector res;
+    auto matched_keys = traverseASTFilter(primary_key, select.where(), res);
+    return std::make_pair(res, !matched_keys);
+}
+
+
+class EmbeddedRocksdbBlockOutputStream : public IBlockOutputStream
+{
+public:
+    explicit EmbeddedRocksdbBlockOutputStream(
+        StorageEmbeddedRocksdb & storage_,
+        const StorageMetadataPtr & metadata_snapshot_)
+        : storage(storage_)
+        , metadata_snapshot(metadata_snapshot_)
+    {}
+
+    Block getHeader() const override { return metadata_snapshot->getSampleBlock(); }
+
+    void write(const Block & block) override
+    {
+        metadata_snapshot->check(block, true);
+        auto rows = block.rows();
+
+        WriteBufferFromOwnString wb_key;
+        WriteBufferFromOwnString wb_value;
+
+        rocksdb::WriteBatch batch;
+        auto columns = metadata_snapshot->getColumns();
+
+        for (size_t i = 0; i < rows; i++)
+        {
+            wb_key.restart();
+            wb_value.restart();
+
+            for (const auto & col : columns)
+            {
+                const auto & type = block.getByName(col.name).type;
+                const auto & column = block.getByName(col.name).column;
+                if (col.name == storage.primary_key)
+                    type->serializeBinary(*column, i, wb_key);
+                else
+                    type->serializeBinary(*column, i, wb_value);
+            }
+            batch.Put(wb_key.str(), wb_value.str());
+        }
+        auto status = storage.rocksdb_ptr->Write(rocksdb::WriteOptions(), &batch);
+        if (!status.ok())
+            throw Exception("Rocksdb write error: " + status.ToString(), ErrorCodes::SYSTEM_ERROR);
+    }
+
+private:
+    StorageEmbeddedRocksdb & storage;
+    StorageMetadataPtr metadata_snapshot;
+};
+
+StorageEmbeddedRocksdb::StorageEmbeddedRocksdb(const StorageID & table_id_,
+        const String & relative_data_path_,
+        const StorageInMemoryMetadata & metadata_,
+        bool attach,
+        Context & context_,
+        const String & primary_key_)
+    : IStorage(table_id_), primary_key{primary_key_}
+{
+    setInMemoryMetadata(metadata_);
+    rocksdb_dir = context_.getPath() + relative_data_path_ + "/rocksdb";
+    if (!attach)
+    {
+        Poco::File(rocksdb_dir).createDirectories();
+    }
+    initDb();
+}
+
+void StorageEmbeddedRocksdb::truncate(const ASTPtr &, const StorageMetadataPtr & , const Context &, TableExclusiveLockHolder &)
+{
+    rocksdb_ptr->Close();
+    Poco::File(rocksdb_dir).remove(true);
+    Poco::File(rocksdb_dir).createDirectories();
+    initDb();
+}
+
+void StorageEmbeddedRocksdb::initDb()
+{
+    rocksdb::Options options;
+    rocksdb::DB * db;
+    options.create_if_missing = true;
```

Comment:
Allow passing extra options will be very useful (since there are tons of them, compression, threads, bloom filters, hash index), and not only rocksdb options but also options for column family, there are API that create options from map:
- `rocksdb::GetDBOptionsFromMap`
- `rocksdb::GetColumnFamilyOptionsFromMap`
- And a separate option for TTL, i.e. `rocsdb::DBWithTTL::Open`.

Plus rocksdb exports tons of statistics:
- `rocksdb::CreateDBStatistics`
- pus there are histograms (`rocksdb::StatsLevel::kExceptHistogramOrTimers`), but they a little bit heavy
worth export it via some metrics or `system.rocksdb`

Reply:
We can split the task to two steps and add support for options in the second step.

Reply:
Ok, since there are lots of options in `rocksdb`, I'd like to continue with options support in another PR.